ui/src/common/common-model.ts

```diff
@@ -16,7 +16,31 @@ export type AngularUnit = 'RADIANS' | 'GRADS';
 
 export type DataType = 'STORED' | 'TEMP';
 
-export type PublishType = 'OFFICIAL' | 'DRAFT';
+export type LayoutBranchId = string;
+export type PublicationState = 'OFFICIAL' | 'DRAFT';
+export type LayoutContext = {
+    publicationState: PublicationState;
+    designId: LayoutBranchId | undefined;
+};
+
+export const officialLayoutContext = (layoutContext: LayoutContext): LayoutContext => ({
+    publicationState: 'OFFICIAL',
+    designId: layoutContext.designId,
+});
+export const draftLayoutContext = (layoutContext: LayoutContext): LayoutContext => ({
+    publicationState: 'DRAFT',
+    designId: layoutContext.designId,
+});
+
+export const officialMainLayoutContext = (): LayoutContext => ({
+    publicationState: 'OFFICIAL',
+    designId: undefined,
+});
+export const draftMainLayoutContext = (): LayoutContext => ({
+    publicationState: 'DRAFT',
+    designId: undefined,
+});
```

Comment:
Aiemmin meillä oli aika paljon kovakoodattuja `'DRAFT'`/`'OFFICIAL'`-määrityksiä hauissa. Ne on pyritty korvaamaan näillä, joilla saadaan funktiosta riippuen joko halutun `PublicationState`:n versio nykyisestä kontekstista tai ihan kokonaan vaan main-paikannuspohjan haluttu versio. Ideana se, että karttanäkymässä pyritään käyttämään noita pohjakontekstin ottavia `draft/officialLayoutContext`-versioita ja kaikkialla sen ulkopuolella missä suunnitelmat eivät kiinnosta (inframodel-näkymä, julkaisut, tietotuotteet jne.) voidaan käyttää noita `Main`-versioita

Reply:
Olion sisällön muuttaminen reduxissa muuttaa aina myös sen parentin. Vaikka redux toolkit saa sen näyttämään niinkuin sinne sisään setattaisi jotain, niin redux on oikeasti immutable oliopuu -> kaikki parentit muuttuu aina kun mikään lapsi muuttuu. Eli siinä mielessä tuo context-olio vaihtuu joka tapauksessa ja olisi sinällään aika se ja sama vain luoda aina uusi kun mitään kontekstista muutetaan.

Olet kyllä ihan oikeassa että olioiden muokkaus jossain muualla on jonkintasoinen riski, vaikkei sitä kenenkään pitäisikään tehdä. Eikös yksi aika helppo toimiva vaihtoehto olisi että noita muokataan kyllä näillä funkkareilla, mutta ne palauttaa vain saman olion jos haluttu tila vastaakin sisään annettua (eli `publicationState === context.publicationState ? context : { //muutettu olio }`.

Reply:
... ja heti perään, rexusissa settaus sinällään varmaan kyllä tekee tuonkin saman ja sittenhän ongelmaa ei ole. Toinen paikka jossa tuo voi vaikuttaa on depsu-arrayt reactille, mutta niissä kannattanee tehdä jotain muuta kikkaa muutenkin koska muuttuneita olioita voi olla montaakin reittiä

Reply:
Hmm, luin juuri että ts:ssä on myös olemassa `Object.freeze`. Onko kukaan koskaan käyttänyt tuota? Tällaisessa flatissa oliossa se käsittääkseni takaisi jopa tyyppitasolla että olio on immutable, ja silloin meillä voisi olla oikeasti vain yksi instanssi per sisältö missä on aika paljon etuja.

Reply:
Ääh, aivan joo. Toimiipas redux ikävästi näissä, ja onpa ärsyttävää muuttaa näitä rakenteita sen rajoittuneisuuksien/suunnitteluratkaisujen takia. Mutta tutkitaan asiaa, onhan tuollainen tarkastelu ja conditional return kanssa ihan vaihtoehto, mutta tutkaisen myös nuo `Object.freeze()`-polut

Reply:
Päädyin muokkaamaan näitä seuraavasti:
* Tarkastetaan noissa muokkaavissa sisääntuleva `layoutContext`, ja jos se on jo haluttua tyyppiä niin palautetaan olio sellaisenaan takaisin, muuten palautetaan uusi olio siten kuin aiemminkin tällä reviewillä
* `main`-kontekstit luodaan ei-exportattuina ja freezattuina olioina, ja `main`-funktiot palauttavat vain ne. Koska kyseessä on yksitasoinen olio, freezauksen pitäisi pitää huoli siitä että pohjaoliot eivät koskaan pääse muuttumaan